### DISCUSSION TITLE
Per-Project Prefixes in Project Factory?

### DISCUSSION THREAD

**Discussion by robrankin** [GoogleCloudPlatform/cloud-foundation-fabric#2651] [General]
#### TL;DR
We'd like to be able to customise the `prefix` in certain areas of FAST (Network and Project Factory at least) without adopting Tenant Factory.

There are no "tenants" in our organization, at least with any autonomy.  A "tenant" in our world is a customer whom we deploy and manage a product for.

#### Problem:
We don't see any need or benefit at this point to adopting the FAST Tenant Factory approach.  There's  chance we're wrong about this, but it seems to us that the FAST definition of tenant doesn't really fit our organization.

What we'd like to do is just be able to change the `prefix` used for each customers set of resources from `main` to `cst1` and `cst2` (see the below example).

It appears that Project Factory doesn't allow this, as the `prefix` is enforced using `data_overrides`, unless I'm missing something?

I've worked up a set of changes to the Project Factory stage and the Project Factory module that allow a custom prefix per project, and I'm happy to open a PR for them, but wanted to see if there was any appetite for this, as the explicit enforcement of a single prefix in the PF stage seems quite intentional.

#### Example:
We have a number of customers, to whom we provide a set of products/services.  Those services comes in the form of a set of environments arranged along the usual lines of dev/stage/prod, but also separated by product lines.  We may have a customer who takes multiple products, resulting in something like the following, using the `hierarchy` structure in Project Factory:

```
├── cst1
│   ├── _config.yaml
│   └── prd
│       ├── _config.yaml
│       ├── env01
│       │   ├── _config.yaml
│       │   └── ce01.yaml
│       └── env02
│           ├── _config.yaml
│           └── ce01.yaml
├── cst2
│   ├── _config.yaml
│   └── prd
│       ├── _config.yaml
│       ├── env01
│       │   ├── _config.yaml
│       │   └── ce01.yaml
│       └── env02
│           ├── _config.yaml
│           └── ce01.yaml
└── main
    ├── _config.yaml
    └── dev
        ├── _config.yaml
        ├── env01
        │   ├── _config.yaml
        │   └── ce01.yaml
        └── env02
            ├── _config.yaml
            └── ce01.yaml
```

The "owning" company here is just "main" and thats the default prefix used across the organization and FAST.

`cst1` and `cst2` are the customers, and they've each taken a 2 "products", here called `env01` and `env02`.

The "products" deployed to those environments are developed in the `main` owners `dev` type environments, but deployed to the customers `prd` type envs.

In each customers `ce01.yaml` project definition the only real change is adding a `prefix` key:

```yaml
data/hierarchy/cst1/prd/env01/ce01.yaml

name: prd-env01-ce01
prefix: cst1
```

The resulting set of projects would be something like:

```
  ~ projects         = {
      + cst1-prd-env01-ce01 = {
          + number     = (known after apply)
          + project_id = "cst1-prd-env01-ce01"
        }
      + cst1-prd-env02-ce01 = {
          + number     = (known after apply)
          + project_id = "cst1-prd-env02-ce01"
        }
      + cst2-prd-env01-ce01 = {
          + number     = (known after apply)
          + project_id = "cst2-prd-env01-ce01"
        }
      + cst2-prd-env02-ce01 = {
          + number     = (known after apply)
          + project_id = "cst2-prd-env02-ce01"
        }
      + main-dev-env01-ce01 = {
          + number     = (known after apply)
          + project_id = "main-dev-env01-ce01"
        }
      + main-dev-env02-ce01 = {
          + number     = (known after apply)
          + project_id = "main-dev-env02-ce01"
        }
        # (1 unchanged attribute hidden)
    }
```

**Comment by ludoo** [comment]
Agreed, tenants are probably overkill. You should already be able to do what you need: just change the project factory stage so `prefix` is not enforced via overrides, but passed is as default (or not even that). At that point, you should be able to declare customized prefixes for each project.

**Reply by ludoo** [reply]
What I forgot to say is that the project factory stage is intended more as a "working example" than a catch-all for every usecase. The code is fairly simple and concise, and overrides/defaults can be easily tweaked.

**Reply by robrankin** [reply]
Yup, I've done that for ourselves, but really the question is wether I should PR back?  It seems like something that'd be useful for everyone else in a future release, but just not sure if the single prefix enforcement is done currently for a specific reason and you didn't want to change that?

**Reply by robrankin** [reply]
Can see the changes here: https://github.com/robrankin/cloud-foundation-fabric/tree/robrankin/per_project_prefix

**Reply by ludoo** [reply]
I'm struggling to find the changes :)

If I understand your point, you would like to have configurable prefixes at the folder level, which trickle down to projects in the folder hierarchy?

It's not a bad idea, but it would incentivize placing project definitions within folders, which we support in the project factory but discourage from doing as you risk getting into a really messy place if a folder is accidentally deleted (been there). Which is why we support and suggest using a flat hierarchy for projects.

What I would probably do in your case is add an extra filesystem-level folders in the (flat) project hierarchy, capture it from the project factory locals, and (optionally) use it in the prefix for projects. I can probably sketch an implementation to explain it better if you want. It's something I don't think the pf should support out of the box, but a nice customization example we should capture in a doc.

**Reply by robrankin** [reply]
Prob shoulda created a PR on my repo, just to highlight the changes 🤦🏻 

https://github.com/robrankin/cloud-foundation-fabric/pull/1/files

As you said, moving the prefix from the data_override to the data_defaults goes most of the way to solving my issue, and I'd already done that... but ran into a couple other things the prompted me to make a few more changes.

In the flat projects folder approach its the potentially large number of files all in a single folder.  To me it felt like a bit of a management pain, so decided to look at the folder hierarchy (with your well documented warning acknowledged).

With the hierarchy folder approach the issue I ran into was that you couldn't use the same filename in multiple levels of the hierarchy, as it caused duplicate key errors in various TF maps, as the keys are built from the basename of the filename (sans the extension).  In the example case I'd like to be able to just copy a file called `ce01.yaml` into multiple customer products/environments, and have PF pick that up whenever we ran a plan/apply.  That way we have an easy(ish) way to add a new project to a load of customers at once, or an even easier way to create a new customer + environments (duplicating whole swathes of the hierarchy en-masse)

The changes in the PF module are all about allowing "duplicate" project files in the folder hierarchy, but having unique map keys.  The changes dont go any further, such as building the project name *from* the directory hierarchy, although for my use case I might actually implement that (have done similar with other TF codebases in the past).

If I understand your final suggestion, your saying to build a folder hierarchy in the filesystem to organize my projects, but don't use that filesystem folder structure for building GCP side folders?

**Reply by ludoo** [reply]
I can see how flat keys from the project folder would be a problem for you, we do this as they make sense when you need to use the project as a replacement.

> In the example case I'd like to be able to just copy a file called ce01.yaml into multiple customer products/environments, and have PF pick that up whenever we ran a plan/apply. 

This makes sense, thanks for explaining it. I see how using the hierarchy to store project files is the sensible approach in your case. I'm not against adding an optional prefix for projects in the folder config files, let's try and settle on how this would work.

This is the high level approach I would like to keep in mind (subject to discussion ofc)

- files should be portable, we need to be able to release sample files that adapt to every user, and support cases where there's multiple orgs sharing the same set of base files (e.g. prod and nonprod orgs)
- the current approach of using overrides/defaults/file local prefix should keep working as is
- folder-defined prefixes should be additive vs the "base prefix" or we would lose portability and risk deviating from the org-wide naming convention
- in a folder hierarchy, we can set the "composed in prefix" but lower level folders and individual projects can change/reset it at will (i.e. it's a default value not a forced one)

My preference would be to add an extra optional attribute (`group_prefix` or somesuch) to folder configs, which when defined gets tacked on the base prefix (whatever comes from overrides/defaults/project file).

As for project keys, I'm pretty sure that projects coming from the hierarchy files keep their full path as key. If that is not the case, it's something we might want to fix. Then we'd have

- a flat hierarchy with file name ---> key
- a nested hierarchy with file path ---> key

We might want to look into supporting a third mode, where the flat hierarchy allows grouping project files in a single folder level, and that folder becomes part of the project key. The folder could additionally allow setting defaults/overrides for the projects contained within (maybe, we have to look at how complex/fragile the implementation is).

- `tenant-a`
  - `foo` (key is `tenant-a/foo`)
- `tenant-b`
  - `foo` (key is `tenant-b/foo`)

This would allow to map your case on either the hierarchy or the flat files.

**Reply by robrankin** [reply]
Regarding your approach, I'm not 100% clear on how it would work.  If I'm reading it correctly, we could end up with project names such as:

* `<global prefix>-<project name>`
* `<global prefix>-<group prefix>-<project name>`

Is that correct?  If so, I'd have two concerns:

1. Names might more easily get close to the length limits.
2. Personally, and possibly being overly pedantic, I much prefer my naming to be of a consistent length and grouping (across the same type of resources).  I find it much easier to scan across long lists of resources if the naming convention used always follows the same pattern.  This is one of the main drivers behind wanting to be able to change prefixes on a per-project basis, so I will always have:

```
<prefix>-<project name>
```

i.e.

```
main-dev-env01-ce01
cst1-stg-env01-ce01
cst1-prd-env01-ce01
cst2-stg-env01-ce01
cst2-prd-env01-ce01
```

Makes listing, sorting and grouping large lists more manageable, and allows people to learn the naming convention and therefore predict resource names without much effort.

> As for project keys, I'm pretty sure that projects coming from the hierarchy files keep their full path as key. If that is not the case, it's something we might want to fix

IIRC line 23 of factory-projects.tf strips off the path using `basename()`, and the result of that determines the map key.

https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/modules/project-factory/factory-projects.tf#L23

**Reply by ludoo** [reply]
So in your case, projects created for tenants use a prefix that maps to the tenant (potentially "locking" a project id they might be interested in themselves in their org), and not to your org/prefix?

This seems slightly strange and I'm not sure it's a model I'd want to have directly supported in the module. :)

The projects key issue we can address ASAP while we keep discussing, feel free to send a PR or log an issue and I'll tackle it as soon as I have some spare time.

**Reply by robrankin** [reply]
> So in your case, projects created for tenants use a prefix that maps to the tenant (potentially "locking" a project id they might be interested in themselves in their org), and not to your org/prefix?

I suppose thats one way of looking at it, but given the allowed prefix length (9?) plus the possible combinations across the rest of the project ID, it seems a little unlikely there would be a problem in reality.

Personally I prefer to have all parts of a resouce name have meaning, thats easy to convey and understand, and having the same prefix across all of them uses valuable naming space without conveying much.  Given those resources exist within a GCP Org, the ultimate ownership and responsibility is already implied, so IMO having the same prefix everywhere doesn't add much.  Using something like a prefix to identify different groups adds meaning.

All that said, I get your point, and being reluctant to add it to the codebase.  I'd just leave one last arguement that moving the current system to "default plus customisable where the user requires it" adds flexibility without (I think?) a lot of complexity, at least from the POV of the users of this codebase.

> The projects key issue we can address ASAP while we keep discussing, feel free to send a PR or log an issue and I'll tackle it as soon as I have some spare time.

I'll see bout PR'ng what I've done so far, give me a day or three.

**Reply by robrankin** [reply]
Opened a PR for just allowing duplicate filenames in the project hierarchy.  Not sure about Windows though, will need to find a Windows machine to double check with.

**Reply by ludoo** [reply]
> default plus customisable where the user requires it

absolutely, I'm just wary of adding complexity to cover individual corner cases given those can be implemented downstream with few changes to the module (like you did), let's take it one step at a time and see where we end up, if that's ok :)

checking the PR

**Reply by robrankin** [reply]
Tested on Windows, and `fileset()` returns forward slash `/` as the path separator so the change in the PR shouldnt cause any platform related issues.

**Comment by jackspyder** [comment]
It would also be good if, in the automation block you can set backend storage account location.

**Reply by ludoo** [reply]
> It would also be good if, in the automation block you can set backend storage account location.

can you clarify what you mean with this?
